crates/sui-core/src/transaction_orchestrator.rs

```diff
@@ -329,38 +311,32 @@ where
             }
             e @ Err(SuiError::TransactionInputObjectsErrors { .. }) => {
                 debug!(?tx_digest, "Orchestrator failed to execute transaction optimistically due to missing parents: {:?}", e);
-
-                match node_sync_handle
-                    .handle_parents_request(state.epoch(), std::iter::once(*tx_digest))
-                    .await?
-                    .next()
-                    .instrument(tracing::debug_span!(
-                        "transaction_orchestrator_execute_tx_via_node_sync"
-                    ))
+                // Now we need to wait for the transaction and its parents to be gossiped and executed from data sync
+                if let Err(err) = state
+                    .transaction_manager
+                    .enqueue(vec![tx_cert.clone()])
```

Comment:
I think we can remove `execute_impl` entirely to make this even more concise. See: https://github.com/MystenLabs/sui/pull/6752/files#diff-f12c25716656fb79726beee3513685fb66fa1dcb746df71cf0ed765e7464edd8R192

Reply:
+1. I think it is better to just use `execute_certificate_with_effects()` instead of `execute_impl()`.

Reply:
sounds good.

Btw, can you remind me why fullnode needs to set shared_locks?

Reply:
When executing a certificate containing shared objects, the shared object locks / versions are read from the storage. So no matter where the certificate is executed, the locks / versions need to be set.